### PR TITLE
Refactor HttpServletMapping to avoid tracking servletPath separately

### LINKED ISSUE #4741: getHttpServletMapping for async dispatch
Unlike the path methods (contextPath, servletPath, pathInfo), the new 4.0 getHttpServletMapping() method does not change when an AsyncContext.dispatch() is done, making it more like include.

### PR DESCRIPTION
**Pull request by gregw**
This completes the refactoring started in #4851 for #4741, by using
the HttpServletMapping field to avoid having the servletPath field
in the Request and instead have a pathInContext field.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

### CODE REVIEW

**Comment by gregw** [comment]
@janbartel @sbordet @lachlan-roberts This is ready for review now.  The CI failure was a flaky one.... rebuilding.

The substance of this change is to avoid the request holding:
```java
   Context context;
   String contextPath;
   String servletPath;
   String pathInfo;
   HttpServletMapping mapping;
```

This has been reduced to 
```java
   Context context;
   String contextPath;
   String pathInContext;
   HttpServletMapping mapping;
```

The mapping itself provides the servletPath and pathInfo (with those being pre-computed for exact and root matches).  
Previously, while there was no context set, the pathInfo was equal to the pathInContext.  Now we just return pathInContext
for the pathInfo when there is no context, but we keep the pathInContext as that can be generally useful and reduces the times
we have to do `pathInContext = addPaths(servletPath, pathInfo)`.

It would have been really nice if we could have reduced down to:
```java
   Context context;
   String pathInContext;
   HttpServletMapping mapping;
```
and just got the context path from the Context object.... but with cross context includes possible (0.0000001% of cases), the contextPath can be for the source context and the context the actual context.   I'm wondering if we should just wrap the context in that case????... but that would be another PR.

**Review by lachlan-roberts — DISMISSED** [review]
LGTM

**Review by janbartel — DISMISSED** [review]
I'm not in love with the method name setContextPaths(String,String) because of the getContextPath() method (ie returns a singular path), but I can't really think of a better one .... setPaths(String,String)??? ... setContextualPaths(String,String)???? They all suck to some degree. Other than that minor quibble, LGTM.

**Comment by gregw** [comment]
I don't like the `setContextPaths(String,String)` setter either - neither the name nor the double barrelled set.
However, if I go ahead and do the context wrapping to get rid of the contextPath field, then this method will change anyway.... so I'm kind of OK with it as I'm keen to do the next PR anyway.

**Comment by gregw** [comment]
@janbartel @lachlan-roberts thanks for the reviews... but looking at it I decided I didn't want to merge with the TODOs and contextPath.  I thought of a neat way to support the cross context include 0.000001% without wrapping, so I've done another commit.    I'll dismiss your reviews so you can review the last commit.

**Comment by gregw** [comment]
Moved back to a draft as Jan suggested offline a more uniform approach.

**Comment by gregw** [comment]
@janbartel @sbordet @lachlan-roberts more javadoc added

**Comment by janbartel** [comment]
@gregw you need to merge from jetty-10.0.x

**Comment by joakime** [comment]
Don't forget to set the Project on PRs

File: jetty-server/src/main/java/org/eclipse/jetty/server/Dispatcher.java [hunk 1]
```diff
@@ -90,6 +90,8 @@ public void include(ServletRequest request, ServletResponse response) throws Ser
         final DispatcherType old_type = baseRequest.getDispatcherType();
         final Attributes old_attr = baseRequest.getAttributes();
         final MultiMap<String> old_query_params = baseRequest.getQueryParameters();
+        final ContextHandler.Context old_context = baseRequest.getContext();
+        final ServletPathMapping old_mapping = baseRequest.getServletPathMapping();
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/Dispatcher.java:94]: All these variable names should break CheckStyle, we don't use snake_case.

File: jetty-server/src/main/java/org/eclipse/jetty/server/Dispatcher.java [hunk 2]
```diff
@@ -136,11 +145,10 @@ protected void forward(ServletRequest request, ServletResponse response, Dispatc
...
 
         final HttpURI old_uri = baseRequest.getHttpURI();
-        final String old_context_path = baseRequest.getContextPath();
-        final String old_servlet_path = baseRequest.getServletPath();
-        final String old_path_info = baseRequest.getPathInfo();
+        final ContextHandler.Context old_context = baseRequest.getContext();
+        final String old_path_in_context = baseRequest.getPathInContext();
         final ServletPathMapping old_mapping = baseRequest.getServletPathMapping();
-
+        final ServletPathMapping source_mapping = baseRequest.findServletPathMapping();
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/Dispatcher.java:151]: Don't use snake_case in variable names.

**Review by sbordet — CHANGES_REQUESTED** [review]
LGTM, but fix the snake_case.

**Review by sbordet — APPROVED** [review]

**Pull request merged**